runtime/parachains/src/hrmp/benchmarking.rs

```diff
@@ -0,0 +1,294 @@
+	}
+
+	hrmp_close_channel {
+		let [(sender, sender_origin), (recipient, _)] =
+			establish_para_connection::<T>(1, 2, ParachainSetupStep::Established);
+		let channel_id = HrmpChannelId { sender, recipient };
+	}: _(sender_origin, channel_id.clone())
+	verify {
+		assert_last_event::<T>(Event::<T>::ChannelClosed(sender, channel_id).into());
+	}
+
+	// NOTE: a single parachain should have the maximum number of allowed ingress and egress
+	// channels.
+	force_clean_hrmp {
+		// ingress channels to a single leaving parachain that need to be closed.
+		let i in 0 .. (<T as configuration::Config>::HrmpMaxInboundChannelsBound::get() - 1);
+		// egress channels to a single leaving parachain that need to be closed.
+		let e in 0 .. (<T as configuration::Config>::HrmpMaxOutboundChannelsBound::get() - 1);
+
+		// we use 10_000 in this benchmark as the prefix of accounts. The components must be greater
+		// than the configured bounds to keep accounts sane.
+		assert!(<T as configuration::Config>::HrmpMaxInboundChannelsBound::get() < PREFIX_0);
+		assert!(<T as configuration::Config>::HrmpMaxOutboundChannelsBound::get() < PREFIX_0);
+
+		// first, update the configs to support this many open channels...
+		assert_ok!(
+			Configuration::<T>::set_hrmp_max_parachain_outbound_channels(frame_system::RawOrigin::Root.into(), e + 1)
+		);
+		assert_ok!(
+			Configuration::<T>::set_hrmp_max_parachain_inbound_channels(frame_system::RawOrigin::Root.into(), i + 1)
+		);
+		// .. and enact it.
+		Configuration::<T>::initializer_on_new_session(&Shared::<T>::scheduled_session());
+
+		let config = Configuration::<T>::config();
+		let deposit: BalanceOf<T> = config.hrmp_sender_deposit.unique_saturated_into();
+
+		let para: ParaId = 1u32.into();
+		let para_origin: crate::Origin = 1u32.into();
+		register_parachain_with_balance::<T>(para, deposit);
+		T::Currency::make_free_balance_be(&para.into_account(), deposit * 256u32.into());
+
+		for ingress_para_id in 0..i {
+			// establish ingress channels to `para`.
+			let ingress_para_id = ingress_para_id + PREFIX_0;
+			let _ = establish_para_connection::<T>(ingress_para_id, 1, ParachainSetupStep::Established);
+		}
+
+		// nothing should be left unprocessed.
+		assert_eq!(HrmpOpenChannelRequestsList::<T>::decode_len().unwrap_or_default(), 0);
+
+		for egress_para_id in 0..e {
+			// establish egress channels to `para`.
+			let egress_para_id = egress_para_id + PREFIX_1;
+			let _ = establish_para_connection::<T>(1, egress_para_id, ParachainSetupStep::Established);
+		}
+
+		// nothing should be left unprocessed.
+		assert_eq!(HrmpOpenChannelRequestsList::<T>::decode_len().unwrap_or_default(), 0);
+
+		// all in all, we have created this many channels.
+		assert_eq!(HrmpChannels::<T>::iter().count() as u32, i + e);
+	}: _(frame_system::Origin::<T>::Root, para, i, e) verify {
+		// all in all, all of them must be gone by now.
+		assert_eq!(HrmpChannels::<T>::iter().count() as u32, 0);
```

Comment:
Here and for other verify:

There is a handy function `assert_storage_consistency_exhaustive` that fails in case any of the storage invariants are invalidated. Really useful to check if something is wrong. 

This is especially important since you are doing shenanigans like `process_hrmp_open_channel_requests` outside of the session change, which should be fine but still not expected use-case.

Reply:
https://github.com/paritytech/polkadot/pull/3876/commits/a47ea77b84e8b54727a4b8a05d838f596b1f05df

Reply:
NOTE: what re-using this (useful) function from the tests will restrict us to only call it in `verify` block of when we are testing benchmarks, which I think is enough, because the tests will run on the (low, mid, high of all benchmarks). Nonetheless, I made it reusable in https://github.com/paritytech/polkadot/pull/3876/commits/745e146f7df89ce7f9c738003b71a1c2c531057e, which we can decide to keep or revert.

Reply:
That makes perfect sense and in fact I also considered the same.